# ICT 782 - Day 7 Notes

# Data Aggregation and Group Operations

When data has been cleaned and transformed, it usually follows that we want to perform some analysis on the cleaned data. We've been using the term 'analysis' rather loosely over the last few weeks, and now we'll give it a clear definition.

> For this course, **analysis** refers to a detailed investigation into the underlying structure and components of a dataset. Through an analysis, we hope to discover some unseen information contained in the data.

The simplest analysis that can be completed on a dataset is computing its summary statistics, such as its mean, quantiles, and variance. This task becomes simplified using methods of the pandas `DataFrame` object. Suppose we have a dataset containing multiple columns, where some columns contain repeated values. Such a dataset is presented below.

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.DataFrame(np.random.rand(30).reshape((15,2)), columns = ['value1','value2'])
data['prov'] = ['AB','AB','BC','SK','MB','ON','QC','AB','NB','NS','QC','BC','ON','QC','AB']
data['party'] = ['L','NDP','C','C','L','NDP','C','L','NDP','L','C','C','L','NDP','C']
data

value1    value2 prov party
0   0.695509  0.415222   AB     L
1   0.746495  0.566775   AB   NDP
2   0.471735  0.419450   BC     C
3   0.107347  0.306446   SK     C
4   0.293298  0.589128   MB     L
5   0.282684  0.366866   ON   NDP
6   0.114019  0.384523   QC     C
7   0.740337  0.468621   AB     L
8   0.181824  0.043087   NB   NDP
9   0.687659  0.552752   NS     L
10  0.707871  0.943667   QC     C
11  0.122998  0.758581   BC     C
12  0.872081  0.204776   ON     L
13  0.183454  0.387120   QC   NDP
14  0.092665  0.515897   AB     C

We can see the summary statistics of the dataset using the `describe()` method. The output of the `describe()` method is another `DataFrame` object.

In [3]:
data.describe()

value1     value2
count  15.000000  15.000000
mean    0.419998   0.461527
std     0.290444   0.215273
min     0.092665   0.043087
25%     0.152411   0.375695
50%     0.293298   0.419450
75%     0.701690   0.559763
max     0.872081   0.943667

However, in the event we want to see the mean for each value of `prov` separately, we can use the `groupby(<key>)` method as follows.

In [4]:
data.groupby('prov').mean()

value1    value2
prov                    
AB    0.568751  0.491629
BC    0.297367  0.589015
MB    0.293298  0.589128
NB    0.181824  0.043087
NS    0.687659  0.552752
ON    0.577382  0.285821
QC    0.335115  0.571770
SK    0.107347  0.306446

The `groupby()` method has placed all common values of `prov` into separate groups inside a `DataFrameGroupBy` object. We then called the `mean()` method on each of these groups.

The full documentation for the `DataFrameGroupBy` object is found in `pandas.core.groupby.generic.DataFrameGroupBy`. Rather than digging through the documentation, we will instead remember that these objects contain groups of values which have been grouped by a given key.

Let's take the dataset from above and take the mean of the various values of `party`.

In [5]:
data.groupby('party').mean()

value1    value2
party                    
C      0.269439  0.554761
L      0.657777  0.446100
NDP    0.348614  0.340962

We can also group a dataset by more than one variable, creating a hierarchical index in our resulting group operation.

In [6]:
data.groupby(['prov','party']).mean()

value1    value2
prov party                    
AB   C      0.092665  0.515897
     L      0.717923  0.441922
     NDP    0.746495  0.566775
BC   C      0.297367  0.589015
MB   L      0.293298  0.589128
NB   NDP    0.181824  0.043087
NS   L      0.687659  0.552752
ON   L      0.872081  0.204776
     NDP    0.282684  0.366866
QC   C      0.410945  0.664095
     NDP    0.183454  0.387120
SK   C      0.107347  0.306446

To see exactly which groups are in the resulting object, we use the fact that grouped datasets are iterables.

In [7]:
for group in data.groupby('prov'):
    print(group)
    print()

('AB',       value1    value2 prov party
0   0.695509  0.415222   AB     L
1   0.746495  0.566775   AB   NDP
7   0.740337  0.468621   AB     L
14  0.092665  0.515897   AB     C)

('BC',       value1    value2 prov party
2   0.471735  0.419450   BC     C
11  0.122998  0.758581   BC     C)

('MB',      value1    value2 prov party
4  0.293298  0.589128   MB     L)

('NB',      value1    value2 prov party
8  0.181824  0.043087   NB   NDP)

('NS',      value1    value2 prov party
9  0.687659  0.552752   NS     L)

('ON',       value1    value2 prov party
5   0.282684  0.366866   ON   NDP
12  0.872081  0.204776   ON     L)

('QC',       value1    value2 prov party
6   0.114019  0.384523   QC     C
10  0.707871  0.943667   QC     C
13  0.183454  0.387120   QC   NDP)

('SK',      value1    value2 prov party
3  0.107347  0.306446   SK     C)



In [8]:
for name, group in data.groupby('prov'):
    print(name)
    print(group)
    print()

AB
      value1    value2 prov party
0   0.695509  0.415222   AB     L
1   0.746495  0.566775   AB   NDP
7   0.740337  0.468621   AB     L
14  0.092665  0.515897   AB     C

BC
      value1    value2 prov party
2   0.471735  0.419450   BC     C
11  0.122998  0.758581   BC     C

MB
     value1    value2 prov party
4  0.293298  0.589128   MB     L

NB
     value1    value2 prov party
8  0.181824  0.043087   NB   NDP

NS
     value1    value2 prov party
9  0.687659  0.552752   NS     L

ON
      value1    value2 prov party
5   0.282684  0.366866   ON   NDP
12  0.872081  0.204776   ON     L

QC
      value1    value2 prov party
6   0.114019  0.384523   QC     C
10  0.707871  0.943667   QC     C
13  0.183454  0.387120   QC   NDP

SK
     value1    value2 prov party
3  0.107347  0.306446   SK     C



Thus, we see that each `name` corresponds to a unique value of `prov`, and each `group` is a dataset corresponding to `name`. The same holds for grouping by more than one key, only now the `name` is a tuple which represents the levels of each `group`.

In [9]:
for name, group in data.groupby(['prov','party']):
    print(name)
    print(group)
    print()

('AB', 'C')
      value1    value2 prov party
14  0.092665  0.515897   AB     C

('AB', 'L')
     value1    value2 prov party
0  0.695509  0.415222   AB     L
7  0.740337  0.468621   AB     L

('AB', 'NDP')
     value1    value2 prov party
1  0.746495  0.566775   AB   NDP

('BC', 'C')
      value1    value2 prov party
2   0.471735  0.419450   BC     C
11  0.122998  0.758581   BC     C

('MB', 'L')
     value1    value2 prov party
4  0.293298  0.589128   MB     L

('NB', 'NDP')
     value1    value2 prov party
8  0.181824  0.043087   NB   NDP

('NS', 'L')
     value1    value2 prov party
9  0.687659  0.552752   NS     L

('ON', 'L')
      value1    value2 prov party
12  0.872081  0.204776   ON     L

('ON', 'NDP')
     value1    value2 prov party
5  0.282684  0.366866   ON   NDP

('QC', 'C')
      value1    value2 prov party
6   0.114019  0.384523   QC     C
10  0.707871  0.943667   QC     C

('QC', 'NDP')
      value1   value2 prov party
13  0.183454  0.38712   QC   NDP

('SK', 'C')
  

We may also use key indexing to get values of a particular column of each group.

In [14]:
data.groupby('prov')['value1'].mean()

prov
AB    0.568751
BC    0.297367
MB    0.293298
NB    0.181824
NS    0.687659
ON    0.577382
QC    0.335115
SK    0.107347
Name: value1, dtype: float64

In [13]:
# To get the means of 'value1' grouped by each province as an array
data.groupby('prov')['value1'].mean().values

array([0.56875123, 0.29736682, 0.29329789, 0.181824  , 0.68765851,
       0.57738247, 0.33511466, 0.10734709])

## Data Aggregation - the `agg()` method

Many functions take in an array of values and return a single number, such as finding the mean, median, standard deviation, or variance. We'll call these 'aggregate' functions. The `agg()` method for pandas objects allows us to apply these functions to subsets of a dataset.

For example, we can use the `agg()` method together with the `groupby()` method to apply aggregate functions to grouped data.

In [15]:
data.groupby('prov').agg('mean')

value1    value2
prov                    
AB    0.568751  0.491629
BC    0.297367  0.589015
MB    0.293298  0.589128
NB    0.181824  0.043087
NS    0.687659  0.552752
ON    0.577382  0.285821
QC    0.335115  0.571770
SK    0.107347  0.306446

Note that this does the same thing as calling `data.groupby('prov').mean()`. However, we can move far beyond this built-in functionality by defining our own aggregate functions and passing them as arguments to the `agg()` method.

In [16]:
def add_threshold(data, threshold = 0.5):
    """ Only add values above the given threshold. """
    
    return data[data > threshold].sum()

data.groupby('prov').agg(add_threshold)

value1    value2
prov                    
AB    2.182340  1.082671
BC    0.000000  0.758581
MB    0.000000  0.589128
NB    0.000000  0.000000
NS    0.687659  0.552752
ON    0.872081  0.000000
QC    0.707871  0.943667
SK    0.000000  0.000000

The predefined aggregate functions for the `agg()` method are passed in by string argument, and are: `'std'`, `'var'`, `'mean'`, `'median'`, `'min'`, `'max'`, `'count'`, `'sum'`, `'prod'`, `'first'` (get the first value not `NaN`), and `'last'` (get the last value not `NaN`).

In [17]:
data.groupby('prov').agg('sum')

value1    value2
prov                    
AB    2.275005  1.966514
BC    0.594734  1.178031
MB    0.293298  0.589128
NB    0.181824  0.043087
NS    0.687659  0.552752
ON    1.154765  0.571642
QC    1.005344  1.715310
SK    0.107347  0.306446

We need not only pass a single function to `agg()`. Instead, we can pass a list of functions or strings for the predefined functions above.

In [18]:
data.groupby('party').agg([add_threshold, 'sum', np.mean])

value1                            value2                    
      add_threshold       sum      mean add_threshold       sum      mean
party                                                                    
C          0.707871  1.616635  0.269439      2.218145  3.328564  0.554761
L          2.995586  3.288883  0.657777      1.141880  2.230499  0.446100
NDP        0.746495  1.394456  0.348614      0.566775  1.363848  0.340962

Additionally, we can apply different aggregate functions to different data columns by passing a dictionary. The `key: value` pairs of our dictionary are passed in as `column: function` or `column: [list of functions]`.

In [19]:
data.groupby('prov').agg({'value1': ['sum','mean'], 
                          'value2': [add_threshold, 'median']})

value1                  value2          
           sum      mean add_threshold    median
prov                                            
AB    2.275005  0.568751      1.082671  0.492259
BC    0.594734  0.297367      0.758581  0.589015
MB    0.293298  0.293298      0.589128  0.589128
NB    0.181824  0.181824      0.000000  0.043087
NS    0.687659  0.687659      0.552752  0.552752
ON    1.154765  0.577382      0.000000  0.285821
QC    1.005344  0.335115      0.943667  0.387120
SK    0.107347  0.107347      0.000000  0.306446

In [20]:
data.groupby('party').agg({'value1': 'prod', 
                           'value2': np.max})

value1    value2
party                    
C      0.000047  0.943667
L      0.090567  0.589128
NDP    0.007039  0.566775

# High-dimensional data

Data comes in all shapes and formats, but there are assumptions we can make about the more abstract properties of the data. For example, if we are given a dataset consisting of 10 *columns* made up of real numbers, then we can conclude that the values in each *row* came from a 10-dimensional 'space' of real numbers.

Without getting too technical, the 10-dimensional 'space' of real numbers is the set of all real numbers $x$ that can be expressed as vectors of the form $x = (x_1,x_2,\ldots,x_{10})$, where each entry is a real number. If we take two such representations $x$ and $y$, then adding together $x$ and $y$ element-wise results in a third vector of the same form. Other vector space properties such as associativity, commutativity, and scalar multiplication also must hold, but those aren't necessary for this course.

Another way to think of a 10-dimensional real vector space is as the Cartesian product of the real numbers 10 times. In other words, the 10-dimensional real vector space $V$ could be expressed as $V = \mathbb{R}\times\mathbb{R}\times\cdots\times\mathbb{R}$.

For an example of a dataset that comes from a 10-dimensional real space, think of taking 10 measurements from 50 individual patients. These measurements would all be real numbers, and since each measurement is independent of the others, we could express the column data for a given row as $(h_1,h_2,\ldots,h_{10})$, where $h_i$ refers to the measurement in the $i$-th column.

Why do we care about dimension at all? It turns out that certain analyses become very difficult in high-dimensional spaces. For example, when performing linear regression we want fewer columns than rows in our dataset (more on this next week). Therefore, there are cases where reducing the number of dimensions can really benefit our work. Principal Component Analysis (PCA) is one dimension reduction method.

**Note:** Adding vectors *only* works when adding two vectors of the same dimension. For example, $(x_1,x_2,x_3) + (y_1,y_2,y_3) = (x_1+y_1, x_2+y_2,x_3+y_3)$ makes sense, but $3 + (0, 1, 3)$ is not well-defined. However, we could do the following: $(3,3,3) + (0,1,3) = (3,4,6)$.

## *Example*

In [21]:
# Example of high-dimensional data

pd.DataFrame(np.random.rand(100).reshape(10,10))

0         1         2         3         4         5         6  \
0  0.049832  0.706607  0.228091  0.116916  0.024518  0.947678  0.823358   
1  0.791025  0.173163  0.814082  0.451067  0.274676  0.433888  0.067128   
2  0.803923  0.092530  0.008602  0.182364  0.947879  0.428353  0.211857   
3  0.004885  0.025942  0.483205  0.697344  0.131852  0.305357  0.098701   
4  0.839822  0.437699  0.151515  0.247542  0.642169  0.365241  0.477518   
5  0.245308  0.740301  0.699685  0.404063  0.624477  0.119647  0.119590   
6  0.910848  0.297667  0.490512  0.463152  0.215332  0.888119  0.848238   
7  0.024394  0.236902  0.107937  0.737483  0.148368  0.059837  0.870328   
8  0.046330  0.657363  0.304254  0.551178  0.432646  0.924464  0.511573   
9  0.137229  0.500769  0.931866  0.489121  0.734986  0.511298  0.389930   

          7         8         9  
0  0.494537  0.081747  0.059572  
1  0.549748  0.697848  0.331079  
2  0.108184  0.017358  0.964885  
3  0.533397  0.963103  0.820273  
4  0.260551  0.613657  0.161007  
5  0.531534  0.348180  0.701852  
6  0.813809  0.091162  0.466309  
7  0.624434  0.042629  0.996287  
8  0.628868  0.375222  0.422999  
9  0.777722  0.068446  0.067069

What if we knew that columns 3, 4, 5, and 7 were highly correlated? We might ask if there is a different way to represent our dataset so that columns 3, 4, 5, and 7 were combined into one column.

# Principal Component Analysis (PCA)

Informally, PCA reduces the number of dimensions in a dataset by combining variables that tend to vary together. We might be given a dataset where two or more columns are correlated. We'd like to explain how each column contributes to the variability in the data, but this cross-correlation makes this impossible. By 'grouping' correlated columns together, we can decompose the dataset into a collection of independent columns.

More formally, for a dataset $X$ with $n$ columns (and a single row of data given by $x$), PCA looks for the optimal encoding function $f(x)$ that de-correlates the columns and sends the data to a lower-dimensional space. In practice, $f(x)$ is defined by a matrix multiplication $D^Tx$, where $T$ is the transpose. The matrix $D$ has the requirement that all of its columns are mutually independent and all columns have norm 1.

The 'PCA decomposition' is where we apply the function $f$ to the data, and is given by $f(x) = D^Tx$. To 'reconstruct' the data, we apply the decoding function $g$ given by $g(f(x)) = DD^Tx = r(x)$. **The number of dimensions in the lower dimensional space is given by the number of columns of the $D$ matrix.** 

To actually find the matrix $D$ with $k$ columns where $k < n$, we just need to find the eigenvectors of $X^TX$ that correspond to the $k$ largest eigenvalues. Then, we simply use these eigenvectors as the columns of $D$. The projections of the data onto the eigenvectors are called the 'Principal Components'.

Wait a minute. We **just** need to find eigenvectors? This isn't a course on linear algebra and matrices! Luckily for us, there are functions in sklearn specifically for finding PCA decompositions. The important thing for us to know is that PCA accomplishes two things:

1. PCA decorrelates the data. If we use all principal components, then we get a decorrelated version of the data.
2. PCA reduces the number of dimensions in the dataset depending on the number of principal components we use.

## Implementing PCA

The two main methods of performing PCA on a dataset are with sklearn and statsmodels. We'll demonstrate both methods here. Our dataset will be called simply `X`. This dataset is created with 4 independent columns, `x1`, `x2`, `x5`, and `x7`. The remaining columns `x3`, `x4`, and `x6` depend on these independent columns. 

Using PCA, we'll hopefully reduce the number of columns in our data and still explain most of the variability in the dataset.

In [22]:
from statsmodels.multivariate.pca import PCA as sPCA
from sklearn.decomposition import PCA as kPCA

In [27]:
X = pd.DataFrame(np.random.rand(60*2).reshape(60,2), columns = ['x1', 'x2'])
X['x5'] = np.random.binomial(1, p = 0.6, size = 60)
X['x7'] = np.random.gamma(shape = 2, size = 60)

X['x3'] = X['x2'] + 5*np.random.randn(60)
X['x4'] = X['x1'] + 1.1*X['x2'] + 0.34
X['x6'] = X['x1'] + np.random.chisquare(7, size = 60)

X = X.reindex(['x1','x2','x3','x4','x5','x6','x7'], axis = 1)
X

x1        x2         x3        x4  x5         x6        x7
0   0.481608  0.825490  -0.126411  1.729647   1   4.848589  1.868805
1   0.583694  0.999886  12.463052  2.023569   1   6.344690  5.738730
2   0.593419  0.478043  -4.339129  1.459267   1   8.202568  4.637713
3   0.007120  0.158745  -4.265457  0.521739   0   8.337400  4.242276
4   0.767789  0.772471  -4.129365  1.957507   1  10.384258  3.457145
5   0.701178  0.844813   2.066119  1.970472   1   7.386470  0.699742
6   0.592857  0.120620   1.333532  1.065540   1   6.345810  1.526139
7   0.394879  0.125302  -2.216341  0.872712   0   4.310181  2.392378
8   0.074051  0.798145  -1.593626  1.292010   0   4.298354  0.741805
9   0.037965  0.485118   0.810167  0.911595   0   3.916466  1.212779
10  0.403325  0.254967  -4.330805  1.023788   1  11.050678  1.933424
11  0.016604  0.639111   0.272479  1.059626   0   8.184784  2.806077
12  0.652392  0.325766  -1.342247  1.350735   1   1.178427  9.633907
13  0.118409  0.581893  -0.591825  1.098491   1   6.648758  0.782205
14  0.101457  0.851576   2.642974  1.378190   1   5.983213  0.981448
15  0.242707  0.616393   0.238261  1.260740   0   8.712868  1.406939
16  0.058410  0.114467  -5.444217  0.524324   1   7.950532  2.333417
17  0.961613  0.691328   0.801494  2.062074   1   9.404334  1.462217
18  0.900126  0.352649   1.708890  1.628041   1   4.623414  1.091739
19  0.584311  0.452229   2.311307  1.421763   0   5.338126  4.117717
20  0.646127  0.444315  -4.115547  1.474874   1   5.331796  2.766026
21  0.796140  0.512785   0.170975  1.700203   0   9.824825  1.029846
22  0.764017  0.811992   2.025160  1.997209   0  17.578885  1.009894
23  0.716370  0.044671  -0.698251  1.105507   1  11.401362  1.884418
24  0.734689  0.904350  -5.705517  2.069475   0  11.171340  5.615080
25  0.433245  0.865744  -4.035890  1.725564   0   2.216952  0.084822
26  0.116699  0.869565   1.077359  1.413221   0  11.958480  1.505743
27  0.497979  0.790221  -2.082226  1.707222   1   7.756217  3.270321
28  0.374147  0.315876   4.576192  1.061611   1   5.990529  0.716430
29  0.023664  0.882774   4.184998  1.334716   0   9.028611  1.799497
30  0.391461  0.195108 -10.591397  0.946080   1   5.169508  1.138588
31  0.857341  0.153774   0.501045  1.366493   0   8.076802  1.555859
32  0.662257  0.372040   4.497449  1.411501   1   1.958897  1.020485
33  0.882806  0.452162   5.214905  1.720184   0  11.089213  1.242309
34  0.836006  0.633704  -5.840201  1.873081   1   8.340353  0.597814
35  0.918779  0.300453  -1.420145  1.589277   0   6.329703  1.320365
36  0.089201  0.193724   4.621181  0.642297   1   5.041188  3.516931
37  0.201411  0.289081  -4.674393  0.859400   0   4.809313  0.974741
38  0.996397  0.475081   2.549884  1.858986   0   4.183346  5.096521
39  0.556128  0.224481  -4.341268  1.143058   0   4.884551  1.818760
40  0.691240  0.764182   7.661738  1.871840   1   6.776673  1.275114
41  0.216785  0.648185  -0.047452  1.269788   1  13.124705  1.847135
42  0.275994  0.288804  -0.799742  0.933678   0   2.820908  0.826128
43  0.781797  0.504716  -4.014033  1.676984   1  12.905912  0.194295
44  0.877898  0.035505   0.170130  1.256954   0   4.274494  1.482252
45  0.818260  0.162117  -4.852714  1.336589   0  16.343676  2.119163
46  0.383773  0.995162  -6.429936  1.818451   1   9.490054  0.511439
47  0.914471  0.093149  -4.601628  1.356935   1   7.211496  0.670037
48  0.152679  0.147171 -14.952803  0.654567   0   8.314450  2.534733
49  0.344092  0.080199   6.170033  0.772310   1   2.999642  1.882439
50  0.650732  0.320239  -1.950562  1.342995   1   7.566375  1.375934
51  0.554371  0.327022   7.170499  1.254095   1   6.402787  4.565419
52  0.242929  0.761371   5.138372  1.420437   1   5.267098  2.912998
53  0.099134  0.878948  -3.688465  1.405977   1   6.862393  4.345411
54  0.245583  0.722374   5.561225  1.380194   1   3.698164  1.457192
55  0.777273  0.856608   5.169094  2.059542   0   6.824802  3.896199
56  0.432449  0.058001   1.454209  0.836250   1   5.515569  0.748309
57  0.50

### *Example 1:* Using sklearn to perform PCA

In [28]:
N = 2

In [32]:
pca1 = kPCA(n_components = N)
pca1.fit(X)

D1 = pca1.components_

print(pca1.explained_variance_ratio_)
print(D1)
print(pca1.transform(X))

[0.6505595  0.26168078]
[[-0.00242758 -0.00910695 -0.96279224 -0.01244522 -0.0107681   0.26677181
  -0.03878442]
 [ 0.01852118  0.02255887  0.26797646  0.04333593 -0.00477578  0.95914131
  -0.07405282]]
[[ -0.93232108  -2.26389974]
 [-12.80982766   2.27673565]
 [  3.91726548  -0.39839895]
 [  3.92440805  -0.27396529]
 [  4.33380164   1.8692492 ]
 [ -2.32460111   0.85934402]
 [ -1.91082245  -0.45386787]
 [  0.94392834  -3.42889576]
 [  0.39467821  -3.12372952]
 [ -2.03214603  -2.90494617]
 [  4.78184318   2.5084066 ]
 [ -0.44078486   0.93639105]
 [ -1.03313242  -6.70947813]
 [  0.04910876  -0.62111118]
 [ -3.25650256  -0.389477  ]
 [ -0.21554198   1.54972923]
 [  5.0196154   -0.82426292]
 [ -0.59866736   2.40473857]
 [ -2.72471157  -1.93783144]
 [ -3.21821541  -1.32273649]
 [  3.00873845  -2.9524648 ]
 [  0.15463472   2.65310441]
 [  0.43243545  10.60772448]
 [  1.38004036   3.82642585]
 [  5.98584039   2.05398935]
 [  2.209409    -5.69894001]
 [ -0.16531113   4.89025659]
 [  1.6725991 

### *Example 2:* Using statsmodels to perform PCA

In [30]:
pca2 = sPCA(X, ncomp = N, demean = False, standardize = False, normalize = False)
D2 = pca2.loadings

print(pca2.rsquare)
print(D2)
print(pca2.factors)

ncomp
0    0.000000
1    0.723287
2    0.955226
Name: rsquare, dtype: float64
      comp_0    comp_1
x1 -0.056063  0.020250
x2 -0.054194  0.027128
x3  0.134744  0.986773
x4 -0.152619  0.060195
x5 -0.063506  0.028021
x6 -0.945773  0.101054
x7 -0.232274  0.102535
       comp_0     comp_1
0   -5.435994   0.721131
1   -6.113523  13.716547
2   -9.765056  -2.836459
3   -9.534036  -2.895673
4  -11.627734  -2.488541
5   -7.319392   3.040719
6   -6.442396   2.221083
7   -5.092901  -1.442234
8   -4.696894  -0.961197
9   -4.044166   1.388379
10 -11.740256  -2.853831
11  -8.553300   1.465157
12  -3.856972  -0.086221
13  -6.818980   0.280417
14  -5.856285   3.449410
15  -8.774514   1.357364
16  -8.947976  -4.265645
17  -9.595601   2.081539
18  -4.777573   2.419256
19  -5.967914   3.352065
20  -6.588596  -3.096758
21  -9.840133   1.399521
22 -16.978984   4.036056
23 -11.589701   0.766640
24 -13.044618  -3.761418
25  -2.994810  -3.613648
26 -11.783940   2.536971
27  -8.770606  -0.773258
28  -5.479097

## Interpreting PCA results

The above output is a little overwhelming, so let's look at it more closely to demystify it. We know that using fewer principal components than the total number of columns will reduce the dimensions of the dataset. Both methods compute the same thing, but use different algorithms so they get different results.

In Example 1 using sklearn, we obtained 3 important pieces of information from the PCA. These are listed with their interpretation as follows:

* `pca1.explained_variance_ratio_` - This is a list of percentages. The list has as many elements as the number of principal components (PCs). Each element represents the percentage of variability in the data for which the corresponding PC accounts. Thus, the first PC accounts for ~52.9% of the variability in the data, and the second PC accounts for ~42.5% of the variability in the data.

* `pca1.components_` - In the above example, we assigned this to the variable `D1`. This assignment to illustrate that the data in `pca1.components_` is exactly the $D$ matrix that encodes the data (it corresponds exactly to the first 2 eigenvectors of $X^TX$).

* `pca1.transform(X)` - This line applies the transformation $D^Tx$ for every row of data $x$ in $X$ and gives us the resulting transformed dataset.

Example 2 gave us exactly the same results as follows:

* `pca2.rsquare` - This gives us the percentage of variability in the data explained by the various PCs. In this case, using 0 PCs explains 0% of the variability, using 1 PC explains ~76.2% of the variability, and using 2 PCs explains ~96.7% of the variability.

* `pca2.loadings` - This was assigned to the variable `D2` to indicate that this is exactly the $D$ matrix that encodes the data (it is presented in the form $D^T$).

* `pca2.factors` - This is the transformed dataset. In other words, the dataset after being reduced to 2 dimensions.

We can make nicer printouts to better show the information given by each method.

In [33]:
sk = 'Using sklearn with {} principal components.'
st = 'Using statsmodels with {} principal components.'

print(sk.format(N))
print('-'*len(sk))
print('PCA explained {:.4f} % of the variability in the data.\n'.format(100*sum(pca1.explained_variance_ratio_)))

print(st.format(N))
print('-'*len(st))
print('PCA explained {:.4f} % of the variability in the data.\n'.format(100*pca2.rsquare[N]))

Using sklearn with 2 principal components.
-------------------------------------------
PCA explained 91.2240 % of the variability in the data.

Using statsmodels with 2 principal components.
-----------------------------------------------
PCA explained 95.5226 % of the variability in the data.



# *Exercises*

1. For the dataset below, display the means for the data grouped by gender.

In [41]:
genders = ['M','F','Other']

data = pd.DataFrame(np.random.rand(400).reshape(100,4), columns = ['mr','tr','hr','kr'])
data['Gender'] = np.random.choice(genders, size = 100)
data

mr        tr        hr        kr Gender
0   0.875749  0.612653  0.857444  0.153430  Other
1   0.189151  0.582665  0.975245  0.492728  Other
2   0.450764  0.999592  0.514956  0.637830  Other
3   0.902150  0.103691  0.451088  0.931128  Other
4   0.808729  0.005555  0.830099  0.677275      M
..       ...       ...       ...       ...    ...
95  0.161751  0.245080  0.457856  0.400155  Other
96  0.745813  0.333426  0.548517  0.102394      F
97  0.337873  0.810050  0.858249  0.773518  Other
98  0.589496  0.435078  0.943750  0.975109      M
99  0.963057  0.926799  0.854730  0.909778      F

[100 rows x 5 columns]

In [43]:
data.groupby('Gender') .agg('mean')

mr        tr        hr        kr
Gender                                        
F       0.506811  0.514329  0.579113  0.467665
M       0.416566  0.560185  0.484907  0.526102
Other   0.521920  0.524769  0.556261  0.521362

In [44]:
# Alternatively, use the following
data.groupby('Gender').mean()

mr        tr        hr        kr
Gender                                        
F       0.506811  0.514329  0.579113  0.467665
M       0.416566  0.560185  0.484907  0.526102
Other   0.521920  0.524769  0.556261  0.521362

2. For the dataset in Exercise 1, apply different aggregate functions to each column. Group the results by gender. Compute the mean for the `mr` column, the median for the `tr` column, the variance for the `hr` column, and count the number of elements between 0.49 and 0.61 in the `kr` column.

In [59]:
def count_between(data, lower = 0.49, upper = 0.61):
    """ Count the number of elements between 0.49 and 0.61 in 'data'. """
    
    d_upper = data[data < upper]
    d_lower = d_upper[d_upper > lower]
    
    return len(d_lower.index)
    
data.groupby('Gender').agg({'mr': ['mean'], 
                            'tr': ['median'],
                            'hr': ['var'],
                            'kr': [count_between]})

mr        tr        hr            kr
            mean    median       var count_between
Gender                                            
F       0.506811  0.479919  0.078785           2.0
M       0.416566  0.622140  0.117787           5.0
Other   0.521920  0.582665  0.088259           5.0

3. This exercise uses the 'wine' dataset that comes with sklearn. Grouping the data by class, compute the mean, median, standard deviation, and variance for only the `flavanoids` column. **Hint:** Computing aggregate functions on grouped data returns a `DataFrame`, so how could you display just the `flavanoids` column?

In [60]:
from sklearn.datasets import load_wine

dat2 = load_wine()
wine = pd.DataFrame(dat2['data'], columns = dat2['feature_names'])
wine['class'] = dat2['target']
wine

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  class  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [61]:
wine.groupby('class').agg({'flavanoids': ['mean','median','std','var']})

flavanoids                           
            mean median       std       var
class                                      
0       2.982373  2.980  0.397494  0.158001
1       2.080845  2.030  0.705701  0.498014
2       0.781458  0.685  0.293504  0.086145

In [62]:
# Equivalently...
wine.groupby('class')['flavanoids'].agg(['mean','median','std','var'])

mean  median       std       var
class                                      
0      2.982373   2.980  0.397494  0.158001
1      2.080845   2.030  0.705701  0.498014
2      0.781458   0.685  0.293504  0.086145

4. What is the dimension of the dataset below?

|Age|RHR|HGH|Sys|
|---|---|---|---|
|14|60|0.34|120|
|20|75|0.45|122|
|15|61|0.23|140|
|16|72|0.97|99|
|21|59|0.55|104|

5. This exercise uses the 'breast cancer Wisconsin' dataset that comes with sklearn. Perform PCA on the dataset using either sklearn or statsmodels. Compare several trials, where the number of components is 2, 3, 5, 10, and 15.

In [63]:
from sklearn.datasets import load_breast_cancer

dat1 = load_breast_cancer()
breast_cancer = pd.DataFrame(dat1['data'], columns = dat1['feature_names'])
breast_cancer

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [64]:
from sklearn.decomposition import PCA

In [68]:
def display_PCA_results(model, N):
    msg = 'Using sklearn with {} principal components.'

    print(msg.format(N))
    print('-'*len(msg))
    print('PCA explained {:.8f} % of the variability in the data.\n'.format(100*sum(model.explained_variance_ratio_)))

In [72]:
trials = [2,3,5,10,15]

for N in trials:
    model = PCA(n_components = N)
    model.fit(breast_cancer)
    
    display_PCA_results(model, N)

Using sklearn with 2 principal components.
-------------------------------------------
PCA explained 99.82211614 % of the variability in the data.

Using sklearn with 3 principal components.
-------------------------------------------
PCA explained 99.97786721 % of the variability in the data.

Using sklearn with 5 principal components.
-------------------------------------------
PCA explained 99.99878765 % of the variability in the data.

Using sklearn with 10 principal components.
-------------------------------------------
PCA explained 99.99998947 % of the variability in the data.

Using sklearn with 15 principal components.
-------------------------------------------
PCA explained 99.99999959 % of the variability in the data.



6. Perform PCA on the 'wine' dataset. Find how many principal components you must use to explain at least 99.999% of the variability in the dataset.

In [78]:
variability = 0.1
N = 0

while variability < 0.99999:
    model = PCA(n_components = N)
    model.fit(wine)
    
    display_PCA_results(model, N)
    variability = sum(model.explained_variance_ratio_)
    N += 1

Using sklearn with 0 principal components.
-------------------------------------------
PCA explained 0.00000000 % of the variability in the data.

Using sklearn with 1 principal components.
-------------------------------------------
PCA explained 99.80876254 % of the variability in the data.

Using sklearn with 2 principal components.
-------------------------------------------
PCA explained 99.98235439 % of the variability in the data.

Using sklearn with 3 principal components.
-------------------------------------------
PCA explained 99.99194388 % of the variability in the data.

Using sklearn with 4 principal components.
-------------------------------------------
PCA explained 99.99707589 % of the variability in the data.

Using sklearn with 5 principal components.
-------------------------------------------
PCA explained 99.99836954 % of the variability in the data.

Using sklearn with 6 principal components.
-------------------------------------------
PCA explained 99.99924302 

In [40]:
# Imagine we have a dataset of several columns from various processes (temp, psi, etc).
# We also have a column indicating upsets.

processes = pd.DataFrame(np.random.rand(20*5).reshape(20,5), columns = ['hrs','temp','psi','conn','lim'])
processes['upset'] = np.random.binomial(1, p=0.3, size = 20)
processes

hrs      temp       psi      conn       lim  upset
0   0.657541  0.754390  0.645277  0.240367  0.762341      1
1   0.445210  0.485293  0.936780  0.505300  0.144348      0
2   0.616643  0.263437  0.140670  0.009850  0.680210      0
3   0.202644  0.175283  0.527156  0.291055  0.798969      0
4   0.322946  0.917184  0.534322  0.533640  0.457390      0
5   0.272710  0.389099  0.271872  0.257171  0.921593      0
6   0.832571  0.846364  0.752737  0.901462  0.414898      0
7   0.971709  0.066641  0.199126  0.448738  0.512977      0
8   0.166028  0.100775  0.633214  0.062031  0.143182      0
9   0.982272  0.084962  0.861538  0.433670  0.928038      1
10  0.680771  0.419049  0.598263  0.871875  0.048526      1
11  0.958898  0.482836  0.111082  0.101798  0.528668      0
12  0.488023  0.293292  0.555190  0.108083  0.110495      0
13  0.685178  0.837979  0.278992  0.532375  0.373965      0
14  0.241655  0.400769  0.344511  0.711973  0.153603      1
15  0.206295  0.398575  0.425113  0.839454  0.574259      0
16  0.920010  0.345182  0.204174  0.596793  0.336555      1
17  0.492053  0.706968  0.417895  0.093395  0.942926      0
18  0.653659  0.313633  0.818189  0.410909  0.902564      1
19  0.473528  0.936252  0.254826  0.254296  0.992086      0